cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -154,59 +155,48 @@ public void updateAppSpec(String namespaceId, String appId, ApplicationSpecifica
     write(key, updated);
   }
 
-  private void updateWorkflowRecord(Id.Program program, String pid, Map<String, String> systemArgs,
-                                    ProgramRunStatus status, @Nullable Throwable failureCause) {
-
-    String workflowNodeId = systemArgs.get(ProgramOptionConstants.WORKFLOW_NODE_ID);
-    String workflowName = systemArgs.get(ProgramOptionConstants.WORKFLOW_NAME);
-    String workflowRun = systemArgs.get(ProgramOptionConstants.WORKFLOW_RUN_ID);
-
-    ProgramRunId workflowRunId = new ProgramRunId(program.getNamespaceId(), program.getApplicationId(),
-                                                  ProgramType.WORKFLOW, workflowName, workflowRun);
+  public Map<String, WorkflowNodeStateDetail> getWorkflowNodeStates(ProgramRunId workflowRunId) {
+    MDSKey key = getProgramKeyBuilder(TYPE_WORKFLOW_NODE_STATE, workflowRunId.getParent().toId())
+      .add(workflowRunId.getRun()).build();
 
-    // Get the run record of the Workflow which started this program
-    MDSKey key = getWorkflowRunRecordKey(workflowRunId.getParent().toId(), workflowRunId.getRun());
+    List<WorkflowNodeStateDetail> nodeStateList = list(key, WorkflowNodeStateDetail.class);
```

Comment:
lets not include the data structure name in the variable. call it ``nodeStateDetails``?